File: shared/src/main/java/io/github/robertomike/hefesto/actions/wheres/Where.java
package io.github.robertomike.hefesto.actions.wheres;

import io.github.robertomike.hefesto.enums.Operator;
import io.github.robertomike.hefesto.enums.WhereOperator;
import lombok.*;

@Getter
@Setter
@ToString
@EqualsAndHashCode(callSuper = true)
@AllArgsConstructor
@RequiredArgsConstructor
public class Where extends BaseWhere {
    private final String field;
    protected Operator operator = Operator.EQUAL;
    protected Object value;

    public Where(String field, Operator operator) {
        this.field = field;
        this.operator = operator;
    }
    public Where(String field, Object value) {
        this.field = field;
        this.value = value;
    }

    public Where(String field, Operator operator, Object value, WhereOperator whereOperator) {
        this.field = field;
        this.operator = operator;
        this.value = value;
        this.whereOperation = whereOperator;
    }

    public static Where make(String field, Object value, WhereOperator whereOperator) {
        return new Where(field, Operator.EQUAL, value, whereOperator);
    }

    public static Where make(String field, Operator operator, WhereOperator whereOperator) {
        return new Where(field, operator, null, whereOperator);
    }
}

File: shared/src/main/kotlin/io/github/robertomike/hefesto/actions/wheres/Where.kt
package io.github.robertomike.hefesto.actions.wheres

import io.github.robertomike.hefesto.enums.Operator
import io.github.robertomike.hefesto.enums.WhereOperator

/**
 * Represents a standard WHERE condition in a query.
 * 
 * Compares a field against a value using a comparison operator.
 * 
 * @property field the field name to filter on (can include table alias, e.g., "user.name")
 * @property operator the comparison operator (EQUAL, GREATER, LIKE, IN, etc.)
 * @property value the value to compare against (can be null for IS_NULL/IS_NOT_NULL)
 * 
 * Example:
 * ```kotlin
 * Where("name", Operator.EQUAL, "John")           // name = 'John'
 * Where("age", Operator.GREATER, 18)              // age > 18
 * Where("email", Operator.LIKE, "%@example.com")  // email LIKE '%@example.com'
 * Where("status", Operator.IN, listOf("active", "pending"))  // status IN ('active', 'pending')
 * Where("deletedAt", Operator.IS_NULL)            // deletedAt IS NULL
 * ```
 */
open class Where(
    val field: String,
    open var operator: Operator = Operator.EQUAL,
    open var value: Any? = null
) : BaseWhere() {

    constructor(field: String, operator: Operator) : this(field, operator, null)

    constructor(field: String, value: Any?) : this(field, Operator.EQUAL, value)

    constructor(field: String, operator: Operator, value: Any?, whereOperator: WhereOperator) : this(field, operator, value) {
        this.whereOperation = whereOperator
    }

    companion object {
        @JvmStatic
        fun make(field: String, value: Any?, whereOperator: WhereOperator): Where {
            return Where(field, Operator.EQUAL, value, whereOperator)
        }

        @JvmStatic
        fun make(field: String, operator: Operator, whereOperator: WhereOperator): Where {
            return Where(field, operator, null, whereOperator)
        }
    }
}
